tracing-subscriber/src/fmt/timing.rs

```diff
@@ -0,0 +1,155 @@
+
+use crate::{registry::LookupSpan, subscribe::Context, Subscribe};
+
+/// A subscriber that tracks the time spent in each span.
+///
+/// This subscriber records the time spent in each span, storing the timing data in the span's
+/// extensions. The subscriber records the time spent in each span as either "idle" time, when the
+/// span is not executing, or "busy" time, when the span is executing. The subscriber records the
+/// time spent in each span as a [`Timing`] resource, which can be accessed by other subscribers.
+#[derive(Debug, Default)]
+pub struct TimingSubscriber;
+
+/// A resource tracking the idle and busy time spent in each span.
+///
+/// This is used by the [`TimingSubscriber`] to track the time spent in each span.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Timing {
+    state: State,
+    idle: Duration,
+    busy: Duration,
+    last: Instant,
+}
+
+impl Default for Timing {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+enum State {
+    /// The span is closed.
+    ///
+    /// No further timing information will be recorded.
+    Closed,
+    #[default]
+    /// The span is currently idle.
+    ///
+    /// Timing information will be recorded when the span becomes busy or closed.
+    Idle,
+    /// The span is currently busy.
+    ///
+    /// Timing information will be recorded when the span becomes idle or closed.
+    Busy,
+}
+
+impl Timing {
+    /// Create a new `Timing` resource.
+    fn new() -> Self {
+        Self {
+            state: State::Idle,
+            idle: Duration::ZERO,
+            busy: Duration::ZERO,
+            last: Instant::now(),
+        }
+    }
+
+    /// Record that the span has become idle.
+    fn enter(&mut self) {
+        if self.state == State::Busy {
+            return;
+        }
+        let now = Instant::now();
+        self.idle += now.duration_since(self.last);
+        self.last = now;
+        self.state = State::Busy;
+    }
+
+    /// Record that the span has become busy.
+    fn exit(&mut self) {
+        if self.state == State::Idle {
+            return;
+        }
+        let now = Instant::now();
+        self.busy += now.duration_since(self.last);
+        self.last = now;
+        self.state = State::Idle;
+    }
+
+    /// Record that the span has been closed.
+    fn close(&mut self) {
+        match self.state {
+            State::Idle => self.idle += Instant::now().duration_since(self.last),
+            State::Busy => self.busy += Instant::now().duration_since(self.last),
+            State::Closed => {}
+        }
+        self.state = State::Closed;
+    }
+
+    /// Get the idle time spent in this span.
+    pub fn idle(&self) -> Duration {
+        self.idle
+    }
+
+    /// Get the busy time spent in this span.
+    pub fn busy(&self) -> Duration {
+        self.busy
+    }
+
+    /// Get the total time spent in this span.
+    pub fn total(&self) -> Duration {
```

Comment:
I added total as an affordance to make viewing logs easier (and not require the viewer / software where the logs end up to do math to work out the total)